src/consolidation.rs

```diff
@@ -40,21 +40,27 @@ pub fn consolidate_slice<T: Ord, R: Semigroup>(slice: &mut [(T, R)]) -> usize {
 
     // Counts the number of distinct known-non-zero accumulations. Indexes the write location.
     let mut offset = 0;
-    for index in 1 .. slice.len() {
 
-        // The following unsafe block elides various bounds checks, using the reasoning that `offset`
+    // We can first move quickly through any prefix of distinct, non-zero elements
+    while offset + 1 < slice.len() && !(slice[offset].0 == slice[offset+1].0) && !slice[offset].1.is_zero() {
```

Comment:
shouldn't this happen inside the loop so that we skip over all the distinct regions rather than just the first one?